models/utils.py

```diff
@@ -1,6 +1,49 @@
 import tensorflow as tf
 import mesh_tensorflow as mtf
 
+def entmax(x, alpha = 1.3, dim = None, n_iter = 2, ensure_sum_one = True):
+    assert alpha > 1 and alpha < 2, 'alpha must be between 1 and 2'
+
+    _gp = lambda x, alpha: x ** (alpha - 1)
+    _gp_inv = lambda x, alpha: mtf.pow(x, (1 / (alpha - 1)))
+    _p = lambda x, alpha: _gp_inv(mtf.relu(x), alpha)
+
+    dim = x.shape[-1] if dim is None else dim
+    d = dim.size
+
+    x = x * (alpha - 1)
+
+    max_val = mtf.reduce_max(x, reduced_dim = dim)
+
+    tau_lo = max_val - _gp(1, alpha)
+    tau_hi = max_val - _gp(1 / d, alpha)
+
+    f_lo = mtf.reduce_sum(_p(x - tau_lo, alpha), reduced_dim = dim) - 1
+
+    dm = tau_hi - tau_lo
+
+    for _ in range(n_iter):
+        dm = dm / 2
+        tau_m = tau_lo + dm
+        p_m = _p(x - tau_m, alpha)
+        f_m = mtf.reduce_sum(p_m, reduced_dim = dim) - 1
+
+        mask = mtf.greater_equal((f_m * f_lo), 0)
+        tau_lo = mtf.where(mask, tau_m, tau_lo)
+
+    if ensure_sum_one:
+        p_m = p_m / mtf.reduce_sum(p_m, reduced_dim = dim)
+
+    return p_m
+
+def sample_categorical(x, dim = None):
```

Comment:
can't use gumbel trick here, so had to build my own sample categorical function

can someone double check my function here?

Reply:
this logic seems sound to me. Have you tested it out yet?

Reply:
yup, i do so at the end of test_models!